### ISSUE TITLE
Test failures when using clang

### ISSUE THREAD

**Issue by ViralBShah** [JuliaLang/openlibm#10]
A number of tests are off by 1 ulp with clang. This could be because the tests are wrong, or clang is doing something wrong, or because openlibm is doing something wrong.

```
$ ./test-double
testing double (without inline functions)
Failure: Test: sinh (0.7) == 0.75858370183953350346
Result:
 is:          7.58583701839533386391e-01   0x1.8465153d5bdbc0000000p-1
 should be:   7.58583701839533497413e-01   0x1.8465153d5bdbd0000000p-1
 difference:  1.11022302462515654042e-16   0x1.00000000000000000000p-53
 ulp       :  1.0000
 max.ulp   :  0.0000
Maximal error of `sinh'
 is      :  1.0000 ulp
 accepted:  0.0000 ulp
Failure: Test: tanh (0.7) == 0.60436777711716349631
Result:
 is:          6.04367777117163607059e-01   0x1.356fb17af2e920000000p-1
 should be:   6.04367777117163496037e-01   0x1.356fb17af2e910000000p-1
 difference:  1.11022302462515654042e-16   0x1.00000000000000000000p-53
 ulp       :  1.0000
 max.ulp   :  0.0000
Failure: Test: tanh (-0.7) == -0.60436777711716349631
Result:
 is:         -6.04367777117163607059e-01  -0x1.356fb17af2e920000000p-1
 should be:  -6.04367777117163496037e-01  -0x1.356fb17af2e910000000p-1
 difference:  1.11022302462515654042e-16   0x1.00000000000000000000p-53
 ulp       :  1.0000
 max.ulp   :  0.0000
Maximal error of `tanh'
 is      :  1.0000 ulp
 accepted:  0.0000 ulp
Failure: Test: exp (1) == e
Result:
 is:          2.71828182845904553488e+00   0x1.5bf0a8b14576a0000000p+1
 should be:   2.71828182845904509080e+00   0x1.5bf0a8b1457690000000p+1
 difference:  4.44089209850062616169e-16   0x1.00000000000000000000p-51
 ulp       :  1.0000
 max.ulp   :  0.0000
Maximal error of `exp'
 is      :  1.0000 ulp
 accepted:  0.0000 ulp
Failure: Test: expm1 (1) == M_El - 1.0
Result:
 is:          1.71828182845904509080e+00   0x1.b7e151628aed20000000p+0
 should be:   1.71828182845904531284e+00   0x1.b7e151628aed30000000p+0
 difference:  2.22044604925031308085e-16   0x1.00000000000000000000p-52
 ulp       :  1.0000
 max.ulp   :  0.0000
Maximal error of `expm1'
 is      :  1.0000 ulp
 accepted:  0.0000 ulp
Failure: log (0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log (-0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log10 (0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log10 (-0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log1p (-1) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log2 (0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: log2 (-0) == -inf plus division by zero exception: Exception "Divide by zero" not set
Failure: gamma (0) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: gamma (-3) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: lgamma (0) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: lgamma (-3) == inf plus division by zero exception: Exception "Divide by zero" not set
Failure: tgamma (0) == NaN plus invalid exception: Exception "Divide by zero" set
Failure: tgamma (0) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: Test: tgamma (0) == NaN plus invalid exception
Result:
 is:          inf   inf
 should be:  nan  nan
Failure: tgamma (-0) == NaN plus invalid exception: Exception "Divide by zero" set
Failure: tgamma (-0) == NaN plus invalid exception: Exception "Invalid operation" not set
Failure: Test: tgamma (-0) == NaN plus invalid exception
Result:
 is:         -inf  -inf
 should be:  nan  nan
Failure: Test: j1 (8.0) == 0.23463634685391462438
Result:
 is:          2.34636346853914601329e-01   0x1.e0890561861820000000p-3
 should be:   2.34636346853914629085e-01   0x1.e0890561861830000000p-3
 difference:  2.77555756156289135106e-17   0x1.00000000000000000000p-55
 ulp       :  1.0000
 max.ulp   :  0.0000
Failure: Test: jn (1, 8.0) == 0.23463634685391462438
Result:
 is:          2.34636346853914601329e-01   0x1.e0890561861820000000p-3
 should be:   2.34636346853914629085e-01   0x1.e0890561861830000000p-3
 difference:  2.77555756156289135106e-17   0x1.00000000000000000000p-55
 ulp       :  1.0000
 max.ulp   :  0.0000
Failure: Test: jn (10, 0.1) == 0.26905328954342155795e-19
Result:
 is:          2.69053289543421753151e-20   0x1.fc3a2ffb93d410000000p-66
 should be:   2.69053289543421572595e-20   0x1.fc3a2ffb93d3b0000000p-66
 difference:  1.80555932286303361224e-35   0x1.80000000000000000000p-116
 ulp       :  6.0000
 max.ulp   :  5.0000
Maximal error of `jn'
 is      :  6.0000 ulp
 accepted:  5.0000 ulp
Failure: Test: y0 (-1.0) == -inf
Result:
 is:         nan  nan
 should be:  -inf  -inf
Failure: Test: y1 (-1.0) == -inf
Result:
 is:         nan  nan
 should be:  -inf  -inf
Failure: Test: y1 (8.0) == -0.15806046173124749426
Result:
 is:         -1.58060461731247520190e-01  -0x1.43b5340f693370000000p-3
 should be:  -1.58060461731247492434e-01  -0x1.43b5340f693360000000p-3
 difference:  2.77555756156289135106e-17   0x1.00000000000000000000p-55
 ulp       :  1.0000
 max.ulp   :  0.0000
Failure: Test: yn (0, -1.0) == -inf
Result:
 is:         nan  nan
 should be:  -inf  -inf
Failure: Test: yn (1, -1.0) == -inf
Result:
 is:         nan  nan
 should be:  -inf  -inf
Failure: Test: yn (1, 8.0) == -0.15806046173124749426
Result:
 is:         -1.58060461731247520190e-01  -0x1.43b5340f693370000000p-3
 should be:  -1.58060461731247492434e-01  -0x1.43b5340f693360000000p-3
 difference:  2.77555756156289135106e-17   0x1.00000000000000000000p-55
 ulp       :  1.0000
 max.ulp   :  0.0000

Test suite completed:
  1129 test cases plus 944 tests for exception flags executed.
  36 errors occurred.
```

**Comment by ViralBShah** [comment]
Is there a way to make sure that on OS X, these tests are using openlibm and not the system libm?

Cc: @staticfloat @tkelman

**Comment by staticfloat** [comment]
`test-double` and friends aren't linked against `libm`, they're even statically linked to `openlibm`, so I don't think there's any question which algorithms are being called.

**Comment by eschnett** [comment]
openlibm contains several implementations of `exp`; there's a generic one, plus various optimized ones for different architectures. The choice is probably made at openlibm build time. Hence my question -- these are all source files in openlibm.

**Comment by ViralBShah** [comment]
Many of these tests are failing due to 1ulp differences, and others due to exceptions. Do we need to fix these for the next release?

**Comment by nalimilan** [comment]
I don't personally care about clang at the moment, so I'm fine if it's not fixed for the next release.

**Comment by ViralBShah** [comment]
Fair enough - but it matters for OS X and FreeBSD. FWIW, the Apple libm fails even more tests.

**Comment by wdv4758h** [comment]
TL;DR: Clang is innocent ...

----

I just compile the test code with `gcc -O0` to see the assembly. It looks like gcc has done some magic optimization for the test. The test for `sinh(0.7)` has already calculated in the compile time ...

```
f2 0f 10 05 9a 95 00    movsd  0x959a(%rip),%xmm0        # 418490 <_IO_stdin_used+0x7480>
00
48 b8 bd bd d5 53 51    movabs $0x3fe8465153d5bdbd,%rax
46 e8 3f
ba 00 00 00 00          mov    $0x0,%edx
be 00 00 00 00          mov    $0x0,%esi
66 0f ef d2             pxor   %xmm2,%xmm2
66 0f 28 c8             movapd %xmm0,%xmm1
48 89 45 f8             mov    %rax,-0x8(%rbp)
f2 0f 10 45 f8          movsd  -0x8(%rbp),%xmm0
bf e8 6f 41 00          mov    $0x416fe8,%edi
e8 73 34 ff ff          callq  402398 <check_float>
```

The value `0x3fe8465153d5bdbd` is `0.758583701839533497413` in IEEE 754 format.

To build with Clang (`-O0`):

```
f2 0f 10 04 25 d8 18    movsd  0x4118d8,%xmm0
41 00
89 45 ec                mov    %eax,-0x14(%rbp)
e8 2a cf ff ff          callq  401860 <sinh@plt>
48 bf 58 36 41 00 00    movabs $0x413658,%rdi
00 00 00
f2 0f 10 0c 25 18 1a    movsd  0x411a18,%xmm1
41 00
0f 57 d2                xorps  %xmm2,%xmm2
31 c0                   xor    %eax,%eax
89 c6                   mov    %eax,%esi
89 c2                   mov    %eax,%edx
e8 39 c5 00 00          callq  410e90 <check_float>
```

Clang **DO** invoke the function call of `sinh`.

I have modified the test code to make gcc invoke the `sinh` function (`test/libm-test.c`) :

```c
FLOAT data = 0.7L;
check_float ("sinh (0.7) == 0.75858370183953350346",  FUNC(sinh) (data), 0.75858370183953350346L, DELTA1548, 0, 0);
```

Then, gcc also fail the test, the result value is the same with clang.

```
test/test-double
testing double (without inline functions)
Failure: Test: sinh (0.7) == 0.75858370183953350346
Result:
 is:          7.58583701839533386391e-01   0x1.8465153d5bdbc0000000p-1
 should be:   7.58583701839533497413e-01   0x1.8465153d5bdbd0000000p-1
 difference:  1.11022302462515654042e-16   0x1.00000000000000000000p-53
 ulp       :  1.0000
 max.ulp   :  0.0000
Maximal error of `sinh'
 is      :  1.0000 ulp
 accepted:  0.0000 ulp

Test suite completed:
  1108 test cases plus 923 tests for exception flags executed.
  2 errors occurred.
Makefile:32: recipe for target 'test' failed
make: *** [test] Error 1
```

When we truely invoke the `sinh` function in openlibm, the test will fail. That's not a compiler issue.

*Clang is innocent*
*I cann't disable that optimization in gcc, even with `-O0` ...*

**Comment by yuyichao** [comment]
`-fno-builtin`

**Comment by wdv4758h** [comment]
I think we should introduce testing framework (like [`gtest`](https://github.com/google/googletest) or [`DejaGnu`](http://www.gnu.org/software/dejagnu/)).

**Comment by wdv4758h** [comment]
@yuyichao yes, with `-fno-builtin` will make gcc test fail.

**Comment by ViralBShah** [comment]
Added in 4a1b310f9175bdff9d0dc5f1caf8a30f28e32a2f

**Comment by ararslan** [comment]
What's the status of this? I'm still getting testing errors with clang on the v0.5.0 release. I'm trying to put together a Homebrew formula for OpenLibm but I figured the tests should pass on OS X first. 😛

**Comment by iblis17** [comment]
Hi @ararslan 
It seems ... not yet fixed
FYI: [comments of f7a18b5](https://github.com/JuliaLang/openlibm/commit/f7a18b519dc6ba654fce73fdfb554d7781c18733) :smile:

**Comment by ViralBShah** [comment]
These have always been there, and I think they are issues with clang. Not sure what is the best way forward here. See if more tests fail to find regressions or disable tests or something?

**Comment by ararslan** [comment]
@iblis17 Haha yes, I assumed it was not yet fixed, I was just looking for an update on any progress toward a fix or decision. 😄 

@ViralBShah If I were to make a [Homebrew](http://brew.sh) formula for this, what would you recommend for [testing](https://github.com/Homebrew/brew/blob/master/share/doc/homebrew/Formula-Cookbook.md#add-a-test-to-the-formula) if not these built-in tests?

**Comment by staticfloat** [comment]
@ararslan I have a branch with the necessary `ulp` tweaks [over here](https://github.com/JuliaLang/openlibm/tree/sf/ulps).  This eliminates all the warnings about test results being off, but it does not get rid of the missing floating-point exceptions.  Does anyone know why these exceptions aren't being set?  As an example, [this test](https://github.com/JuliaLang/openlibm/blob/5501bf0576491e213d474462a218c3258a70db78/test/libm-test.c#L3173) is failing because [this line](https://github.com/JuliaLang/openlibm/blob/5501bf0576491e213d474462a218c3258a70db78/src/e_logf.c#L46) is failing to set the "divided by zero" exception.

**Comment by simonbyrne** [comment]
Clang/LLVM doesn't support fenv access generally, including  setting/accessing floating point exceptions.

**Comment by ViralBShah** [comment]
@simonbyrne Does building with gcc on homebrew perhaps make more sense?

**Comment by simonbyrne** [comment]
Or just disable the exception tests if built with clang.

**Comment by ararslan** [comment]
@simonbyrne So in making a Homebrew formula for this, I should provide the option to build with GCC or Clang and disable tests if Clang is used? Which should be the default compiler, Clang because it's Mac? It seems like there should be *some* kind of test, if only to ensure that it installed properly.

**Comment by staticfloat** [comment]
I haven't been able to compile `openlibm` with `gcc` on OSX successfully.  The assembler that `gcc` uses is too old to recognize some commands or something like that.  I think @simonbyrne's suggestion of just omitting the exception tests on `clang` is the best path forward.

**Comment by staticfloat** [comment]
I've got a PR ready for review: https://github.com/JuliaLang/openlibm/pull/132

**Comment by certik** [comment]
I tested with the latest Clang 5.0.0 and all tests pass for me.

**Comment by ararslan** [comment]
Yes I think this has been fixed.

**Comment by galderz** [comment]
Any progress on the Homebrew formula for openlibm? The one in [homebrew-juliadeps](https://github.com/staticfloat/homebrew-juliadeps/blob/master/openlibm.rb) is outdated and there are no tests. Is that being maintained? I've created my own formula for 0.7.0 pretty much as a copy of that but no tests. For my own local testing it worked.

**Comment by ViralBShah** [comment]
It would be good for homebrew to have its own formula for openlibm. Given the move to BinaryBuilder, I doubt we will continue maintaining the homebrew-juliadeps.

**Comment by galderz** [comment]
@ViralBShah I agree, and I would have sent a PR for the one I made, but I lack tests for it and that seems to be mandatory requirement. I wondered if any tests have been added/implemented by those that worked in the brew packages mentioned earlier in this issue.

**Comment by ViralBShah** [comment]
This package it self does have tests. Just do `make test`.

**Comment by galderz** [comment]
Oh right, that was easy! 😅

Sent a [PR](https://github.com/Homebrew/homebrew-core/pull/50543) to homebrew-core to add openlibm 0.7.0 🤞
